placement/algo/algo.go

```diff
@@ -74,22 +75,31 @@ func (a rackAwarePlacementAlgorithm) RemoveHost(ps placement.Snapshot, leavingHo
 }
 
 func (a rackAwarePlacementAlgorithm) AddHost(ps placement.Snapshot, addingHost placement.Host) (placement.Snapshot, error) {
-	addingHostShards := placement.NewEmptyHostShardsFromHost(addingHost)
+	var addingHostShards placement.HostShards
+	var err error
+	if addingHostShards, err = getNewHostShardsForSnapshot(ps, addingHost); err != nil {
+		return nil, err
+	}
 	return a.addHostShards(ps, addingHostShards)
 }
 
 func (a rackAwarePlacementAlgorithm) ReplaceHost(ps placement.Snapshot, leavingHost, addingHost placement.Host) (placement.Snapshot, error) {
+	var err error
+	var addingHostShards placement.HostShards
+	if addingHostShards, err = getNewHostShardsForSnapshot(ps, addingHost); err != nil {
+		return nil, err
+	}
+
 	var ph PlacementHelper
 	var leavingHostShards placement.HostShards
-	var err error
 	if ph, leavingHostShards, err = newRemoveHostPlacementHelper(ps, leavingHost); err != nil {
 		return nil, err
 	}
-	addingHostShards := placement.NewEmptyHostShardsFromHost(addingHost)
+
 	var shardsUnassigned []uint32
 	// move shards from leaving host to adding host
 	for _, shard := range leavingHostShards.Shards() {
-		if moved := ph.MoveShard(shard, leavingHostShards, addingHostShards); !moved {
+		if moved := ph.MoveShard(shard, leavingHostShards, addingHostShards, !a.options.LooseRackCheck()); !moved {
```

Comment:
Instead of passing the boolean flag to the `MoveShard` method, I wonder if it makes more sense to pass the options to the PlacementHelper constructor so its `MoveShard` method behaves differently depending on the options. This way we don't have to expose it in the `PlacementHelper` interface, and if we need to add/relax more constraints in the future, it's easy to do so.